engine/src/main/java/org/hibernate/search/engine/common/spi/SearchIntegrationEnvironment.java

```diff
@@ -0,0 +1,168 @@
+import org.hibernate.search.engine.environment.classpath.spi.ClassResolver;
+import org.hibernate.search.engine.environment.classpath.spi.DefaultClassResolver;
+import org.hibernate.search.engine.environment.classpath.spi.DefaultResourceResolver;
+import org.hibernate.search.engine.environment.classpath.spi.DefaultServiceResolver;
+import org.hibernate.search.engine.environment.classpath.spi.ResourceResolver;
+import org.hibernate.search.engine.environment.classpath.spi.ServiceResolver;
+
+/**
+ * The environment of a search integration,
+ * i.e. the configuration properties, classes, resources, services and beans.
+ */
+public final class SearchIntegrationEnvironment implements AutoCloseable {
+	public static final String CONFIGURATION_PROPERTIES_MASK = "hibernate.search";
+
+	private final ConfigurationPropertySource propertySource;
+	private final ConfigurationPropertyChecker propertyChecker;
+	private final ClassResolver classResolver;
+	private final ResourceResolver resourceResolver;
+	private final ServiceResolver serviceResolver;
+	private final BeanResolver beanResolver;
+	private final BeanProvider beanProvider;
+
+	private SearchIntegrationEnvironment(Builder builder) {
+		propertySource = builder.propertySource.withMask( CONFIGURATION_PROPERTIES_MASK );
+		propertyChecker = builder.propertyChecker;
+
+		AggregatedClassLoader aggregatedClassLoader = null;
+
+		if ( builder.classResolver != null ) {
+			classResolver = builder.classResolver;
+		}
+		else {
+			aggregatedClassLoader = AggregatedClassLoader.createDefault();
+			classResolver = DefaultClassResolver.create( aggregatedClassLoader );
+		}
+
+		if ( builder.resourceResolver != null ) {
+			resourceResolver = builder.resourceResolver;
+		}
+		else {
+			if ( aggregatedClassLoader == null ) {
+				aggregatedClassLoader = AggregatedClassLoader.createDefault();
+			}
+			resourceResolver = DefaultResourceResolver.create( aggregatedClassLoader );
+		}
+
+		if ( builder.serviceResolver != null ) {
+			serviceResolver = builder.serviceResolver;
+		}
+		else {
+			if ( aggregatedClassLoader == null ) {
+				aggregatedClassLoader = AggregatedClassLoader.createDefault();
+			}
+			serviceResolver = DefaultServiceResolver.create( aggregatedClassLoader );
+		}
+
+		beanProvider = builder.beanProvider;
+		beanResolver = BeanResolverImpl.create( classResolver, serviceResolver, beanProvider, propertySource );
+	}
+
+	private SearchIntegrationEnvironment(SearchIntegrationEnvironment source, ConfigurationPropertySource propertySource,
+			ConfigurationPropertyChecker checker) {
+		this.propertySource = propertySource.withMask( CONFIGURATION_PROPERTIES_MASK );
+		this.propertyChecker = checker;
+		this.classResolver = source.classResolver;
+		this.resourceResolver = source.resourceResolver;
+		this.serviceResolver = source.serviceResolver;
+		beanProvider = source.beanProvider;
+		beanResolver = source.beanResolver;
+	}
+
+	@Override
+	public void close() {
+		beanProvider.close();
```

Comment:
Is `beanProvider` supposed to be nullable?

Reply:
You are right.